cpp/dolfinx/fem/interpolate.h

```diff
@@ -602,8 +633,217 @@ void interpolate(Function<T>& u, const Function<T>& v,
     // Get mesh and check that functions share the same mesh
     if (mesh != v.function_space()->mesh())
     {
-      throw std::runtime_error(
-          "Interpolation on different meshes not supported (yet).");
+      const int nProcs = dolfinx::MPI::size(MPI_COMM_WORLD);
+      const auto mpi_rank = dolfinx::MPI::rank(MPI_COMM_WORLD);
+
+      const int tdim = u.function_space()->mesh()->topology().dim();
+      const auto cell_map
+          = u.function_space()->mesh()->topology().index_map(tdim);
+      const int num_cells = cell_map->size_local() + cell_map->num_ghosts();
+      std::vector<std::int32_t> cells(num_cells, 0);
+      std::iota(cells.begin(), cells.end(), 0);
+
+      // Collect all the points at which values are needed to define the
+      // interpolating function
+      const xt::xtensor<double, 2> x = fem::interpolation_coords(
+          *u.function_space()->element(), *u.function_space()->mesh(), cells);
+
+      // Create a global bounding-box tree to quickly look for processes that
+      // might be able to evaluate the interpolating function at a given point
+      dolfinx::geometry::BoundingBoxTree bb(*v.function_space()->mesh(), tdim,
+                                            0.0001);
+      auto globalBB = bb.create_global_tree(v.function_space()->mesh()->comm());
+
+      // Get the list of processes that might be able to evaluate each point
+      const auto collisions
+          = dolfinx::geometry::compute_collisions(globalBB, xt::transpose(x));
+
+      // Compute which points need to be sent to which process
+      std::vector<std::vector<double>> pointsToSend(nProcs);
+      // Reserve enough space to avoid the need for reallocating
+      std::for_each(pointsToSend.begin(), pointsToSend.end(),
+                    [&x](auto& el) { el.reserve(3 * x.shape(1)); });
+      // The coordinates of the points that need to be sent to process p go into
+      // pointsToSend[p]
+      for (decltype(x.shape(1)) i = 0; i < x.shape(1); ++i)
+      {
+        for (const auto& p : collisions.links(i))
+        {
+          const auto point = xt::col(x, i);
+          pointsToSend[p].insert(pointsToSend[p].end(), point.cbegin(),
+                                 point.cend());
+        }
+      }
+
+      std::vector<std::int32_t> nPointsToSend(nProcs);
+      std::transform(pointsToSend.cbegin(), pointsToSend.cend(),
+                     nPointsToSend.begin(),
+                     [](const auto& el) { return el.size(); });
+
+      // Share a matrix where element (i, j) is the number of coordinates that
+      // process i sends to process j
+      std::vector<std::int32_t> allPointsToSend(nProcs * nProcs);
+      MPI_Allgather(nPointsToSend.data(), nProcs, MPI_INT32_T,
+                    allPointsToSend.data(), nProcs, MPI_INT32_T,
+                    MPI_COMM_WORLD);
```

Comment:
`MPI_COMM_WORLD` should be replaced with a user defined comm, e.g. from the mesh.

Reply:
Thanks, I changed all occurrences. Should I also check that the two meshes have the same communicator?